### PR TITLE
Cloudstack reconciler validator

### PR DESCRIPTION
**Pull request by ddjjia**
*Issue #, if available:*
- [Cloudstack validator#1055](https://github.com/aws/eks-anywhere-internal/issues/1055)
- [Validate Datacenter Config#1259](https://github.com/aws/eks-anywhere-internal/issues/1259)

*Description of changes:*
Building validator in controller to use cmk to run validations against Cloudstack
https://github.com/aws/eks-anywhere/pull/5452

*Testing (if applicable):*
manually tested: 
- manually add new secret to create new wl cluster
- successfully validate dc
- use invalid secret (invalid api-key) to create new wl cluster
  - "Reconciler error" err=<  getting domain info - 🙈 Error: failed to decode response
- use unexistent secret name to create cluster
  - cloudstack_datacenter_controller.go:73] "Failed to reconcile CloudStackDatacenterConfig" err="Secret \"secret2\" not found" controller="cloudstackdatacenterconfig"
  - Reconciler error" err="Secret \"secret2\" not found"
- fail to validate dc stops reconciliation
  - CloudStackDatacenterConfig hasn't been validated yet"

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: controllers/cloudstack_datacenter_controller.go [hunk 1]
```diff
@@ -36,20 +39,26 @@ func (r *CloudStackDatacenterReconciler) SetupWithManager(mgr ctrl.Manager) erro
 func (r *CloudStackDatacenterReconciler) Reconcile(ctx context.Context, req ctrl.Request) (_ ctrl.Result, reterr error) {
 	log := ctrl.LoggerFrom(ctx)
 
+	// Fetch the VsphereDatacenter object
 	cloudstackDatacenter := &anywherev1.CloudStackDatacenterConfig{}
 	if err := r.client.Get(ctx, req.NamespacedName, cloudstackDatacenter); err != nil {
 		return ctrl.Result{}, fmt.Errorf("failed getting cloudstack datacenter config: %v", err)
 	}
 
+	// Initialize the patch helper
```
[the remark this answers is not in the document]
Comment (ddjjia) [reply controllers/cloudstack_datacenter_controller.go:48]: patch helper is needed to patch datacenterconfig, for this case is StatusObservedGeneration

File: controllers/cloudstack_datacenter_controller.go [hunk 2]
```diff
@@ -36,20 +39,26 @@ func (r *CloudStackDatacenterReconciler) SetupWithManager(mgr ctrl.Manager) erro
...
 	defer func() {
 		// Always attempt to patch the object and status after each reconciliation.
-		if err := patchHelper.Patch(ctx, cloudstackDatacenter); err != nil {
-			log.Error(reterr, "Failed to patch cloudstackDatacenter")
+		patchOpts := []patch.Option{}
+		if reterr == nil {
+			patchOpts = append(patchOpts, patch.WithStatusObservedGeneration{})
+		}
+		if err := patchHelper.Patch(ctx, cloudstackDatacenter, patchOpts...); err != nil {
+			log.Error(reterr, "Failed to patch cloudstackdatacenterconfig")
```
[the remark this answers is not in the document]
Comment (ddjjia) [reply controllers/cloudstack_datacenter_controller.go:61]: same above

File: pkg/providers/cloudstack/reconciler/reconciler.go [hunk 3]
```diff
@@ -58,13 +60,34 @@ func (r *Reconciler) Reconcile(ctx context.Context, log logr.Logger, cluster *an
...
+		clusters.CleanupStatusAfterValidate,
 		r.ReconcileControlPlane,
 		r.CheckControlPlaneReady,
 		r.ReconcileCNI,
 		r.ReconcileWorkers,
 	).Run(ctx, log, clusterSpec)
 }
 
+// ValidateDatacenterConfig updates the cluster status if the CloudStackDatacenter status indicates that the spec is invalid.
+func (r *Reconciler) ValidateDatacenterConfig(ctx context.Context, log logr.Logger, spec *c.Spec) (controller.Result, error) {
```
[the remark this answers is not in the document]
Comment (ddjjia) [reply pkg/providers/cloudstack/reconciler/reconciler.go:73]: error is returned in datacenter controller not here

**Review by cxbrowne1207 — APPROVED** [review]

**Review by vivek-koppuru — COMMENTED** [review]
We should also have an issue for moving the retrieval of the validator to our individual validation functions as a later PR.

File: pkg/providers/cloudstack/reconciler/reconciler.go [hunk 4]
```diff
@@ -58,13 +60,34 @@ func (r *Reconciler) Reconcile(ctx context.Context, log logr.Logger, cluster *an
...
+	log = log.WithValues("phase", "validateDatacenterConfig")
+	log.Info("Validating datacenter config")
+	dataCenterConfig := spec.CloudStackDatacenter
+
+	if dataCenterConfig.Status.SpecValid {
```
[the remark this answers is not in the document]
Comment (ddjjia) [reply pkg/providers/cloudstack/reconciler/reconciler.go:78]: No, validation only happens in the datacenter reconciler.

**Review by vivek-koppuru — APPROVED** [review]

**Pull request merged**